### PR TITLE
Bug 1952174: status: Report old versions while rolling out new

### PR DESCRIPTION
**Pull request by Miciah**
#### `computeOperatorStatusVersions`: Delete receiver

* `pkg/operator/controller/status/controller.go` (`Reconcile`): Delete receiver in call to `computeOperatorStatusVersions`.
(`computeOperatorStatusVersions`): Delete unused receiver.
* `pkg/operator/controller/status/controller_test.go` (`TestComputeOperatorStatusVersions`): Adjust call to `computeOperatorStatusVersions`.

#### status: Report old versions while rolling out new

For each operand, keep reporting the operand's old version in the clusteroperator `status.versions` value until the new version of the operand is fully rolled out, so that the clusteroperator doesn't report the new version before the operand has finished upgrading.  Likewise, keep reporting the operator's old version until all operands are updated to their respective new versions.

* `pkg/operator/controller/status/controller.go` (`Reconcile`): Use the new `operatorState` fields and the new `computeOldVersions`, `computeNewVersions`, and `computeCurrentVersions` functions to compute versions maps.  Pass these versions maps to `computeOperatorProgressingCondition`.  Pass the current versions map to `computeOperatorStatusVersions`.
(`operatorState`): Add `dnsDaemonSet`, `nodeResolverDaemonSet`, `dnsPodList`, and `nodeResolverPodList` fields.
(`getOperatorState`): Initialize the new fields in `operatorState`.
(`computeOperatorStatusVersions`): Rewrite to take the new versions map and return a slice of `OperandVersion`.
(`computeOperatorProgressingCondition`): Replace the `OperandVersion` slice and string parameter values with parameters for the old, new, and current versions maps.  Use the versions maps to compute the `Progressing` status condition.
(`computeOldVersions`): New function.  Convert the `OperandVersion` slice value from the clusteroperator's `status.versions` field into a versions map.
(`computeNewVersions`): New function.  Convert the versions from the reconciler into a versions map.
(`computeCurrentVersions`): New function.  Use the old and new versions maps and dns and node-resolver daemonsets and pods to compute the current versions map.
* `pkg/operator/controller/status/controller_test.go` (`TestComputeOperatorProgressingCondition`): Adjust for changes to `computeOperatorProgressingCondition`.
(`TestComputeOperatorStatusVersions`): Delete test.
(`TestComputeCurrentVersions`): New test.  Verify that `computeCurrentVersions` behaves as expected.
* `go.mod`: Add `k8s.io/kubectl` dependency for the `"k8s.io/kubectl/pkg/util/podutils".IsPodAvailable` function, used in `computeCurrentVersions`.
* `go.sum`:
* `vendor/*`: Regenerate.

---

Follow-up to #269.

### CODE REVIEW

File: pkg/operator/controller/status/controller.go [hunk 1]
```diff
@@ -202,10 +206,34 @@ func initializeClusterOperator(co *configv1.ClusterOperator) {
...
 	haveNamespace bool
-	namespace     corev1.Namespace
-	haveDNS       bool
-	dns           operatorv1.DNS
+
+	// namespace is the operand namespace.
+	namespace corev1.Namespace
+
+	// haveDNS indicates whether the "default" dnses.openshift.operator.io
+	// CRD exists.
```
Comment (sgreene570) [inline pkg/operator/controller/status/controller.go:216]:
```suggestion
	// CR exists.
```
total nit, don't feel obligated to fix this.

File: go.mod [hunk 2]
```diff
@@ -4,14 +4,19 @@ go 1.16
...
+	github.com/json-iterator/go v1.1.11 // indirect
 	github.com/kevinburke/go-bindata v3.11.0+incompatible
 	github.com/openshift/api v0.0.0-20210416094334-c22782737ea0
 	github.com/openshift/build-machinery-go v0.0.0-20210409131504-b1828cc0cdad
 	github.com/sirupsen/logrus v1.6.0
+	golang.org/x/net v0.0.0-20210510120150-4163338589ed // indirect
 	k8s.io/api v0.21.0
 	k8s.io/apimachinery v0.21.0
 	k8s.io/client-go v0.21.0
+	k8s.io/kubectl v0.21.0
```
Comment (sgreene570) [inline go.mod:18]: At first this made me a little nervous! 
But then I remembered that go mod is smart enough to only import what we actually need from kubectl.

**Comment by Miciah** [comment]
Reviewing [the latest run of the e2e-upgrade CI job](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-dns-operator/274/pull-ci-openshift-cluster-dns-operator-master-e2e-upgrade/1395572551454822400), [`e2e-upgrade/openshift-e2e-test/artifacts/junit/e2e-intervals_20210521-033733.json`](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-dns-operator/274/pull-ci-openshift-cluster-dns-operator-master-e2e-upgrade/1395572551454822400/artifacts/e2e-upgrade/openshift-e2e-test/artifacts/junit/e2e-intervals_20210521-033733.json) shows that the upgrade went from 04:29:04 to 04:33:11:

```json
        {
            "level": "Warning",
            "locator": "clusteroperator/dns",
            "message": "condition/Progressing status/True reason/Upgrading kube-rbac-proxy to \"registry.build02.ci.openshift.org/ci-op-d7g7114b/stable@sha256:3869910c1e208b125bdecd4ac2d8b2cae42efe221c704491b86aa9b18ce95a65\".\nUpgrading operator to \"4.8.0-0.ci.test-2021-05-21-025413-ci-op-d7g7114b-latest\".\nUpgrading coredns to \"registry.build02.ci.openshift.org/ci-op-d7g7114b/stable@sha256:bcdefdbcee8af1e634e68a850c52fe1e9cb31364525e30f5b20ee4eacb93c3e8\".\nUpgrading openshift-cli to \"registry.build02.ci.openshift.org/ci-op-d7g7114b/stable@sha256:7c886e097d93dec6633844c20611937b028d283ddcd11fb2c33d89233dd00abe\".",
            "from": "2021-05-21T04:29:04Z",
            "to": "2021-05-21T04:33:11Z"
        },
```

And looking at [`e2e-upgrade/openshift-e2e-test/artifacts/e2e.log`](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-dns-operator/274/pull-ci-openshift-cluster-dns-operator-master-e2e-upgrade/1395572551454822400/artifacts/e2e-upgrade/openshift-e2e-test/artifacts/e2e.log), it looks like the reported version for the node-resolver operand was updated at 04:29:37:

    May 21 04:29:37.557 I clusteroperator/dns versions: openshift-cli registry.build02.ci.openshift.org/ci-op-d7g7114b/stable-initial@sha256:7c886e097d93dec6633844c20611937b028d283ddcd11fb2c33d89233dd00abe -> registry.build02.ci.openshift.org/ci-op-d7g7114b/stable@sha256:7c886e097d93dec6633844c20611937b028d283ddcd11fb2c33d89233dd00abe

And then the reported versions for the coredns and kube-rbac-proxy operands as well as for the operator itself were updated at 04:33:11:

    May 21 04:33:11.128 I clusteroperator/dns versions: kube-rbac-proxy registry.build02.ci.openshift.org/ci-op-d7g7114b/stable-initial@sha256:3869910c1e208b125bdecd4ac2d8b2cae42efe221c704491b86aa9b18ce95a65 -> registry.build02.ci.openshift.org/ci-op-d7g7114b/stable@sha256:3869910c1e208b125bdecd4ac2d8b2cae42efe221c704491b86aa9b18ce95a65, operator 4.8.0-0.ci.test-2021-05-21-025011-ci-op-d7g7114b-initial -> 4.8.0-0.ci.test-2021-05-21-025413-ci-op-d7g7114b-latest, coredns registry.build02.ci.openshift.org/ci-op-d7g7114b/stable-initial@sha256:bcdefdbcee8af1e634e68a850c52fe1e9cb31364525e30f5b20ee4eacb93c3e8 -> registry.build02.ci.openshift.org/ci-op-d7g7114b/stable@sha256:bcdefdbcee8af1e634e68a850c52fe1e9cb31364525e30f5b20ee4eacb93c3e8

So versions were updated at 04:29:37 and 04:33:11, which fall within the period of 04:29:04 to 04:33:11 during which the operator indicated it was progressing, which is as expected.  

I'll push a fix for https://github.com/openshift/cluster-dns-operator/pull/274#discussion_r635588164 and hope for another good CI run.

File: pkg/operator/controller/status/controller.go (hunk 1 above)
Comment (Miciah) [reply pkg/operator/controller/status/controller.go]: Sorry, couldn't help feeling obligated to fix "CRD" here and above, as well as the API group name.

**Comment by sgreene570** [comment]
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-dns-operator/274/pull-ci-openshift-cluster-dns-operator-master-e2e-upgrade/1395628913098493952

failed on the `KubeAPIErrorBudgetBurn` alert
/retest

**Comment by Miciah** [comment]
In [the latest run of the e2e-upgrade CI job](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-dns-operator/274/pull-ci-openshift-cluster-dns-operator-master-e2e-upgrade/1395835728105050112), [`e2e-intervals_20210521-212654.json`](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-dns-operator/274/pull-ci-openshift-cluster-dns-operator-master-e2e-upgrade/1395835728105050112/artifacts/e2e-upgrade/openshift-e2e-test/artifacts/junit/e2e-intervals_20210521-212654.json) shows that the upgrade went from 22:17:33 to 22:21:36:

```json
        {
            "level": "Warning",
            "locator": "clusteroperator/dns",
            "message": "condition/Progressing status/True reason/Upgrading operator to \"4.8.0-0.ci.test-2021-05-21-202215-ci-op-k590kbmz-latest\".\nUpgrading coredns to \"registry.build02.ci.openshift.org/ci-op-k590kbmz/stable@sha256:bcdefdbcee8af1e634e68a850c52fe1e9cb31364525e30f5b20ee4eacb93c3e8\".\nUpgrading openshift-cli to \"registry.build02.ci.openshift.org/ci-op-k590kbmz/stable@sha256:9c0914ba655514ee0a144b255f34c9ed7c5c66a7d90e04024355b9edb080dca9\".\nUpgrading kube-rbac-proxy to \"registry.build02.ci.openshift.org/ci-op-k590kbmz/stable@sha256:3869910c1e208b125bdecd4ac2d8b2cae42efe221c704491b86aa9b18ce95a65\".",
            "from": "2021-05-21T22:17:33Z",
            "to": "2021-05-21T22:21:36Z"
        },
```

And looking at [`e2e.log`](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-dns-operator/274/pull-ci-openshift-cluster-dns-operator-master-e2e-upgrade/1395835728105050112/artifacts/e2e-upgrade/openshift-e2e-test/artifacts/e2e.log), it looks like the reported version for the node-resolver operand was updated at 22:18:24:

    May 21 22:18:24.678 I clusteroperator/dns versions: openshift-cli registry.build02.ci.openshift.org/ci-op-k590kbmz/stable-initial@sha256:9c0914ba655514ee0a144b255f34c9ed7c5c66a7d90e04024355b9edb080dca9 -> registry.build02.ci.openshift.org/ci-op-k590kbmz/stable@sha256:9c0914ba655514ee0a144b255f34c9ed7c5c66a7d90e04024355b9edb080dca9

And then the reported versions for the coredns and kube-rbac-proxy operands as well as for the operator itself were updated at 22:21:36:

    May 21 22:21:36.818 I clusteroperator/dns versions: operator 4.8.0-0.ci.test-2021-05-21-201759-ci-op-k590kbmz-initial -> 4.8.0-0.ci.test-2021-05-21-202215-ci-op-k590kbmz-latest, coredns registry.build02.ci.openshift.org/ci-op-k590kbmz/stable-initial@sha256:bcdefdbcee8af1e634e68a850c52fe1e9cb31364525e30f5b20ee4eacb93c3e8 -> registry.build02.ci.openshift.org/ci-op-k590kbmz/stable@sha256:bcdefdbcee8af1e634e68a850c52fe1e9cb31364525e30f5b20ee4eacb93c3e8, kube-rbac-proxy registry.build02.ci.openshift.org/ci-op-k590kbmz/stable-initial@sha256:3869910c1e208b125bdecd4ac2d8b2cae42efe221c704491b86aa9b18ce95a65 -> registry.build02.ci.openshift.org/ci-op-k590kbmz/stable@sha256:3869910c1e208b125bdecd4ac2d8b2cae42efe221c704491b86aa9b18ce95a65


So versions were updated at 22:18:24 and 22:21:36, which fall within the period of 22:17:33 to 22:21:36 during which the operator indicated it was progressing, which is as expected.

**Comment by sgreene570** [comment]
looks good! Thanks for the in-depth explanation in https://github.com/openshift/cluster-dns-operator/pull/274#issuecomment-847285632

/lgtm

**Pull request merged**